### PR TITLE
Make Domain Routing ISM enumerable

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Converts `DomainRoutingIsm` to use enumerable mapping to facilitate reading actual ISM config from a deployed address. Adds a `DefaultFallbackRoutingIsm` extension to `DomainRoutingIsm` for easy PI composability.

### Drive-by changes

Migrates ICAs to use Router abstraction

### Related issues

https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2576

### Backward compatibility

ICAs not compatible

### Testing

Unit tests

### CODE REVIEW

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 1]
```diff
@@ -473,16 +456,13 @@ contract InterchainAccountRouter is
         bytes32 _router,
         bytes32 _ism
     ) private {
-        require(_router != bytes32(0), "invalid router address");
```
Comment (aroralanuk) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:476]: why did you remove this check, I don't think you're checking for 0 addresses anywhere for isms or routers

**Review by aroralanuk — APPROVED** [review]

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 2]
```diff
@@ -37,27 +34,15 @@ contract InterchainAccountRouter is
...
-    // ============ Private Storage ============
-    uint32[] private _domains;
-
     // ============ Public Storage ============
-    mapping(uint32 => bytes32) public routers;
     mapping(uint32 => bytes32) public isms;
 
     // ============ Upgrade Gap ============
 
-    // gap for upgrade safety
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:49]: redundant with line 40

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:476]: its already checked in Router
